akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedMetadataSpec.scala

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2009-2024 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.scaladsl
+
+import com.typesafe.config.ConfigFactory
+import org.scalatest.wordspec.AnyWordSpecLike
+
+import akka.actor.testkit.typed.scaladsl.LogCapturing
+import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
+import akka.actor.testkit.typed.scaladsl.TestProbe
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.RecoveryCompleted
+import akka.persistence.typed.SnapshotCompleted
+
+object EventSourcedMetadataSpec {
+
+  private val conf = ConfigFactory.parseString(s"""
+      akka.persistence.journal.plugin = "akka.persistence.journal.inmem"
+      akka.persistence.journal.inmem.test-serialization = on
+      akka.persistence.snapshot-store.plugin = "slow-snapshot-store"
+      slow-snapshot-store.class = "${classOf[SlowInMemorySnapshotStore].getName}"
+    """)
+
+}
+
+class EventSourcedMetadataSpec
+    extends ScalaTestWithActorTestKit(EventSourcedMetadataSpec.conf)
+    with AnyWordSpecLike
+    with LogCapturing {
+
+  private def behavior(pid: PersistenceId, probe: ActorRef[String]): Behavior[String] =
+    Behaviors.setup(
+      ctx =>
+        EventSourcedBehavior[String, String, String](pid, "", {
+          (state, command) =>
+            state match {
+              case "stashing" =>
+                command match {
+                  case "unstash" =>
+                    probe ! s"${EventSourcedBehavior.currentMetadata[String](ctx)} unstash"
+                    Effect.persist("normal").persistMetadata("meta-unstashing").thenUnstashAll()
+                  case _ =>
+                    Effect.stash()
+                }
+              case _ =>
+                command match {
+                  case "cmd" =>
+                    probe ! s"${EventSourcedBehavior.currentMetadata[String](ctx)} onCommand"
+                    Effect
+                      .persist("evt")
+                      .persistMetadata("meta")
```

Comment:
At first glance, this was a bit confusing for me. How can I persist multiple events with different metadatas? I would expect to pass event + meta together: `Effect.persist("evt", "meta")` I know that this might be problematic for the current API. Maybe we need some form of an additional wrapper? `EventWithMeta(event, meta)`

Reply:
I noted that in https://github.com/akka/akka/pull/32628#discussion_r1925332880
Probably good enough? Would be more API to add but also more difficult to handle in the internal representation because we don't want to wrap in the normal case when there is no metadata.

Reply:
How about: `Effect.persistWithMeta(event, meta)` ?

Reply:
but it would be for the case when persisting more than one event that it would be needed.

for that it would have to be `Seq[EventWithMeta]` api

Reply:
and it's possible to store more than one metadata entry per event as well
```
Effect
  .persist("evt")
  .persistMetadata("meta1")
  .persistMetadata("meta2")
```

so to fully support that it would be something like:
```
class EventWithMeta[E](event: E, meta: Seq[Any])
```

Reply:
Yes, but having separate methods would minimize the potential conflict with the current API:
```
.persistWithMeta(event, meta) //for a single one
.persistWithMeta(Seq[EventWithMeta]) //for many
```